src/common/thread.pool.hh

```diff
@@ -0,0 +1,50 @@
+#ifndef H_ACQUIRE_STORAGE_ZARR_THREAD_POOL_V0
+#define H_ACQUIRE_STORAGE_ZARR_THREAD_POOL_V0
+
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <optional>
+#include <queue>
+#include <string>
+#include <thread>
+
+namespace acquire::sink::zarr::common {
+struct ThreadPool final
+{
+  public:
+    using JobT = std::function<bool(std::string&)>;
+
+    // The error handler `err` is called when a job returns false. This
+    // can happen when the job encounters an error, or otherwise fails. The
+    // std::string& argument to the error handler is a diagnostic message from
+    // the failing job and is logged to the error stream by the Zarr driver when
+    // the next call to `append()` is made.
+    ThreadPool(size_t n_threads, std::function<void(const std::string&)> err);
+    ~ThreadPool() noexcept;
+
+    void push_to_job_queue(JobT&& job);
+
+    /**
+     * @brief Block until all jobs on the queue have processed, then spin down
+     * the threads.
+     * @note After calling this function, the job queue no longer accepts jobs.
+     */
+    void await_stop() noexcept;
+
+  private:
+    std::function<void(const std::string&)> error_handler_;
+
+    std::vector<std::thread> threads_;
+    mutable std::mutex jobs_mutex_;
+    std::condition_variable cv_;
+    std::queue<JobT> jobs_;
+
+    std::atomic<bool> is_accepting_jobs_;
+
+    std::optional<ThreadPool::JobT> pop_from_job_queue_() noexcept;
```

Comment:
We tend to use `[[nodiscard]]` for error-indicating return values (`bool write()`, for example).

Reply:
That makes sense. Using `[[nodiscard]]` for error handling and resource management is the canonical use case. I personally add it to most functions that return a value unless it's okay for the client to ignore it. A good example of why this broader approach is valuable is `std::vector::empty()`. If you're new to C++, you might mistakenly think it clears the vector. Since the standard library marked it as `[[nodiscard]]`, the intent is much clearer, and it can be enforced by static analyzers.